File: Rust/src/bin/206_reverse_linked_list.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}
 
impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
      let (mut prev,mut cur)=(None,head);
      while let Some(mut node) = cur {
        cur=node.next;
        node.next=prev;
        prev=Some(node);
      }
      prev
    }
}

struct Solution {}

fn main() {
    let values=vec![1,2,3,4,5];
    let mut dummy=ListNode::new(0);
    let mut current=&mut dummy;
    for &val in &values {
        current.next=Some(Box::new(ListNode::new(val)));
        current=current.next.as_mut().unwrap();
    }
    let l1=dummy.next;
    println!("{:?}",l1);
    
    println!("{:?}",Solution::reverse_list(l1));
}

File: Python/206.ReverseLinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def pList(l):
    if not l: 
        print()
        return
    print(l.val, end=' -> ')
    pList(l.next)



class Solution:
    def reverseList(self, head):
        # if head is None: return None
        # nHead=head
        # if head.next:
        #     nHead=self.reverseList(head.next)
        #     head.next.next=head
        # head.next=None
        # return nHead
        prev,cur=None,head
        while cur:
            tmp=cur.next # get next before updating it
            cur.next=prev # reverse
            prev=cur # update prev
            cur=tmp # update cur
        return prev
            
        
s=Solution()
l1=ListNode(1,ListNode(2,ListNode(3,ListNode(4,ListNode(5)))))
pList(l1)
l1_1=s.reverseList(l1)
pList(l1_1)